### PR TITLE
feat(sdk): add `bucket.signedUrl()` for `tf-gcp` platform

### LINKED ISSUE #4599: Implement `cloud.Bucket` inflight method `signedUrl()` for `tf-gcp` target
### Feature Spec

SignedUrl method support for GCP bucket inflight client along with permissions using Google Storage Client library. [Reference](https://cloud.google.com/storage/docs/access-control/signing-urls-with-helpers#client-libraries_1)

- This PR #4376 support all client methods except signedUrl for bucket resource.
- There are few permissions related tasks has to be made before supporting signedUrl.  

### Use Cases

It will make this possible. 
```js
const BUCKET_NAME = "test-bucket";
  const KEY = "test-file-1";
  const VALUE = "test-content-1";

  const storage = new MockStorage();
  await storage.bucket(BUCKET_NAME).put(KEY, VALUE);

  const client = new BucketClient(BUCKET_NAME, storage as any);
  const res = await client.signedUrl(KEY, {       // This will work
    duration: Duration.fromSeconds(86400),
  });
  expect(res).toBe(
    `https://storage.googleapis.com/${BUCKET_NAME}/${KEY}?X-Goog-Algorithm=MOCKED`
  );
```

### Implementation Notes

_No response_

### Component

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.
* If this issue is labeled **needs-discussion**, it means the spec has not been finalized yet. Please reach out on the #dev channel in the [Wing Slack](https://t.winglang.io/slack).

### PR DESCRIPTION
**Pull request by garysassano**
Closes #4599

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingsdk/src/target-tf-gcp/bucket.ts [hunk 1]
```diff
@@ -61,13 +62,26 @@ export class Bucket extends cloud.Bucket {
...
+    // This is disabled by default, but required for generating presigned URLs
+    const iamServiceAccountCredentialsApi = new ProjectService(
+      this,
+      "IamServiceAccountCredentialsApi",
+      {
+        service: "iamcredentials.googleapis.com",
+        disableDependentServices: false,
+        disableOnDestroy: false,
+      }
+    );
```
Comment (tsuf239) [inline libs/wingsdk/src/target-tf-gcp/bucket.ts:75]: Can we add it only if we use signedUrl?

File: examples/tests/sdk_tests/bucket/signed_url.test.w [hunk 2]
```diff
@@ -69,7 +69,12 @@ test "signedUrl PUT" {
 }
 
 test "signedUrl duration option is respected" {
-  let ACCESS_DENIED_ERROR = "<Error><Code>AccessDenied</Code><Message>Request has expired</Message>";
+  let isExpiredTokenError = (output: str) => {
+    let awsError = output.contains("<Code>AccessDenied</Code><Message>Request has expired</Message>");
+    let gcpError = output.contains("<Code>ExpiredToken</Code><Message>Invalid argument.</Message>");
+    return awsError || gcpError;
+  };
```
Comment (tsuf239) [inline examples/tests/sdk_tests/bucket/signed_url.test.w:76]: we can maybe condition it on the `util.env("WING_TARGET")`? (within an object or a switch-case maybe?)

File: libs/wingsdk/src/shared-gcp/bucket.inflight.ts [hunk 3]
```diff
@@ -211,13 +212,47 @@ export class BucketClient implements IBucketClient {
...
+   */
   public async signedUrl(
-    _key: string,
-    _options?: BucketSignedUrlOptions
+    key: string,
+    opts: BucketSignedUrlOptions
   ): Promise<string> {
-    throw new Error("Method not implemented.");
+    const gcsFile = this.bucket.file(key);
+    let gcsAction: "read" | "write" | "delete" | "resumable";
```
Comment (tsuf239) [inline libs/wingsdk/src/shared-gcp/bucket.inflight.ts:227]: is there an enum for that?

File: libs/wingsdk/src/shared-gcp/bucket.inflight.ts [hunk 4]
```diff
@@ -211,13 +212,47 @@ export class BucketClient implements IBucketClient {
...
+    }
+
+    // Generate the presigned URL
+    const [signedUrl] = await gcsFile.getSignedUrl({
+      version: "v4",
+      action: gcsAction,
+      expires: Date.now() + (opts?.duration?.seconds ?? 900) * 1000,
+    });
+
+    return signedUrl;
```
Comment (tsuf239) [inline libs/wingsdk/src/shared-gcp/bucket.inflight.ts:256]: Nice!

File: libs/wingsdk/src/shared-gcp/bucket.inflight.ts (hunk 3 above)
Comment (garysassano) [reply libs/wingsdk/src/shared-gcp/bucket.inflight.ts:227]: I checked for it, but that's just how it's defined sadly.

image

**Review by tsuf239 — APPROVED** [review]
Passed all aws tests as well :) 
https://github.com/winglang/wing/actions/runs/7638382004

**Pull request merged**